File: ios/Foodie/FoodSearch.swift
import Foundation

// MARK: - Models

/// One entry in the bundled offline table (see CommonFoods.swift). Per 100 g.
struct CommonFood: Identifiable, Hashable {
    let id: String
    let name: String
    let kcal: Double
    let protein: Double
    let carbs: Double
    let fat: Double
    let servings: [Serving]
}

/// A named portion, e.g. "1 medium" = 118 g.
struct Serving: Hashable {
    let label: String
    let grams: Double
}

/// A search result from either source, normalised to per-100 g macros so the
/// portion picker can scale everything the same way.
struct FoodHit: Identifiable, Hashable {
    enum Source: Hashable { case common, openFoodFacts }

    let id: String
    let name: String
    let brand: String?
    let kcal: Double      // per 100 g
    let protein: Double   // per 100 g
    let carbs: Double     // per 100 g
    let fat: Double       // per 100 g
    let servings: [Serving]
    let source: Source

    /// Portions offered in the picker: the food's own servings, then 100 g.
    var portionOptions: [Serving] {
        servings + [Serving(label: "100 g", grams: 100)]
    }

    /// Picker text for a serving. Open Food Facts labels often already spell out
    /// the weight ("1 portion (40 g)"), so only append it when it's missing.
    static func portionLabel(_ serving: Serving) -> String {
        let hasWeight = serving.label.range(of: #"\d\s*g\b"#, options: .regularExpression) != nil
        return hasWeight ? serving.label : "\(serving.label) (\(Int(serving.grams.rounded())) g)"
    }

    /// Scale to a gram weight and round into the integer-only FoodEntry shape.
    func entry(grams: Double) -> FoodEntry {
        let k = grams / 100
        let title = [brand, name].compactMap { $0 }.filter { !$0.isEmpty }.joined(separator: " ")
        return FoodEntry(
            name: title.isEmpty ? name : title,
            kcal: Int((kcal * k).rounded()),
            protein: Int((protein * k).rounded()),
            carbs: Int((carbs * k).rounded()),
            fat: Int((fat * k).rounded())
        )
    }
}

// MARK: - Errors

enum FoodSearchError: LocalizedError {
    case notFound
    case noNutrition
    case offline
    case server

    var errorDescription: String? {
        switch self {
        case .notFound:    return "No product found for that barcode."
        case .noNutrition: return "That product has no nutrition data yet."
        case .offline:     return "You're offline — bundled foods still work."
        case .server:      return "Food database is busy. Try again in a moment."
        }
    }
}

// MARK: - Service

/// Looks foods up in the bundled table first, then Open Food Facts.
///
/// Open Food Facts is free, needs no API key and has no monthly quota; it only
/// asks that clients send an identifying User-Agent, which we do below.
actor FoodSearchService {
    static let shared = FoodSearchService()

    private let session: URLSession
    /// Barcode results are stable, so keep them for the life of the process.
    private var barcodeCache: [String: FoodHit] = [:]

    init(session: URLSession? = nil) {
        if let session {
            self.session = session
        } else {
            let config = URLSessionConfiguration.default
            config.timeoutIntervalForRequest = 12
            config.waitsForConnectivity = false
            config.httpAdditionalHeaders = [
                "User-Agent": "Foodie/\(Self.appVersion) (iOS) - https://mcursch.github.io/foodie/"
            ]
            self.session = URLSession(configuration: config)
        }
    }

    private static var appVersion: String {
        Bundle.main.object(forInfoDictionaryKey: "CFBundleShortVersionString") as? String ?? "1.0"
    }

    // MARK: Bundled table

    /// Ranked matches from the offline table. Instant, no network.
    nonisolated func searchCommon(_ query: String, limit: Int = 8) -> [FoodHit] {
        let q = query.folding(options: [.diacriticInsensitive, .caseInsensitive], locale: .current)
            .trimmingCharacters(in: .whitespacesAndNewlines)
        guard q.count >= 2 else { return [] }

        return CommonFoods.all
            .compactMap { food -> (FoodHit, Int)? in
                guard let score = Self.score(name: food.name, query: q) else { return nil }
                return (FoodHit(id: "common:\(food.id)", name: food.name, brand: nil,
                                kcal: food.kcal, protein: food.protein, carbs: food.carbs,
                                fat: food.fat, servings: food.servings, source: .common), score)
            }
            // Ties keep the alphabetical order the generated table is already in.
            .sorted { $0.1 > $1.1 }
            .prefix(limit)
            .map(\.0)
    }

    /// Higher is better; nil means no match. Prefers whole-name prefixes, then
    /// word starts, so "chick" ranks "Chicken breast" above "Chickpeas".
    private static func score(name: String, query: String) -> Int? {
        let n = name.folding(options: [.diacriticInsensitive, .caseInsensitive], locale: .current)
        guard let range = n.range(of: query) else { return nil }
        if range.lowerBound == n.startIndex { return 100 - n.count / 4 }
        let before = n[n.index(before: range.lowerBound)]
        if before == " " || before == "," || before == "-" { return 60 - n.count / 4 }
        return 20 - n.count / 4
    }

    // MARK: Open Food Facts — text search

    /// Branded products matching `query`. Returns [] rather than throwing when
    /// the network or the service is unavailable — bundled results still show.
    func searchProducts(_ query: String, limit: Int = 20) async -> [FoodHit] {
        let q = query.trimmingCharacters(in: .whitespacesAndNewlines)
        guard q.count >= 2 else { return [] }
        guard var comps = URLComponents(string: "https://search.openfoodfacts.org/search") else { return [] }
        comps.queryItems = [
            URLQueryItem(name: "q", value: q),
            URLQueryItem(name: "page_size", value: String(limit)),
            URLQueryItem(name: "fields", value: "code,product_name,brands,nutriments,serving_size"),
        ]
        guard let url = comps.url else { return [] }

        do {
            let (data, response) = try await session.data(from: url)
            guard let http = response as? HTTPURLResponse, http.statusCode == 200 else { return [] }
            let payload = try JSONDecoder().decode(SearchResponse.self, from: data)
            // OFF holds one entry per country/packaging, so a popular product comes
            // back a dozen times. Collapse rows that agree on brand, name and kcal.
            var seen = Set<String>()
            return payload.hits.compactMap { $0.toHit() }.filter { hit in
                let key = "\((hit.brand ?? "").lowercased())|\(hit.name.lowercased())|\(Int(hit.kcal.rounded()))"
                return seen.insert(key).inserted
            }
        } catch {
            return []
        }
    }

    // MARK: Open Food Facts — barcode

    /// Look up a scanned barcode. Throws so the scanner can explain failures.
    func lookup(barcode: String) async throws -> FoodHit {
        let code = barcode.trimmingCharacters(in: .whitespacesAndNewlines)
        if let cached = barcodeCache[code] { return cached }

        guard var comps = URLComponents(string: "https://world.openfoodfacts.org/api/v2/product/\(code).json")
        else { throw FoodSearchError.notFound }
        comps.queryItems = [
            URLQueryItem(name: "fields", value: "code,product_name,brands,nutriments,serving_size")
        ]
        guard let url = comps.url else { throw FoodSearchError.notFound }

        let data: Data
        let response: URLResponse
        do {
            (data, response) = try await session.data(from: url)
        } catch let error as URLError where error.code == .notConnectedToInternet || error.code == .networkConnectionLost {
            throw FoodSearchError.offline
        } catch {
            throw FoodSearchError.server
        }

        guard let http = response as? HTTPURLResponse else { throw FoodSearchError.server }
        if http.statusCode == 404 { throw FoodSearchError.notFound }
        guard http.statusCode == 200 else { throw FoodSearchError.server }

        guard let payload = try? JSONDecoder().decode(ProductResponse.self, from: data),
              payload.status == 1, let product = payload.product
        else { throw FoodSearchError.notFound }

        guard let hit = product.toHit(code: code) else { throw FoodSearchError.noNutrition }
        barcodeCache[code] = hit
        return hit
    }
}

// MARK: - Wire format

/// Open Food Facts returns `brands` as a comma string on the product endpoint
/// but as an array on the search endpoint, so decode either shape.
private struct FlexibleBrands: Decodable {
    let value: String?

    init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()
        if let s = try? container.decode(String.self) {
            value = s.split(separator: ",").first.map { $0.trimmingCharacters(in: .whitespaces) }
        } else if let a = try? container.decode([String].self) {
            value = a.first
        } else {
            value = nil
        }
    }
}

private struct OFFProduct: Decodable {
    let code: String?
    let productName: String?
    let brands: FlexibleBrands?
    let nutriments: [String: OFFValue]?
    let servingSize: String?

    enum CodingKeys: String, CodingKey {
        case code
        case productName = "product_name"
        case brands
        case nutriments
        case servingSize = "serving_size"
    }

    /// Build a hit, or nil when the product has no usable calorie figure.
    func toHit(code overrideCode: String? = nil) -> FoodHit? {
        let id = overrideCode ?? code ?? UUID().uuidString
        let name = (productName ?? "").trimmingCharacters(in: .whitespacesAndNewlines)
        guard !name.isEmpty else { return nil }

        // Prefer the explicit kcal field; fall back to converting kJ.
        let kcal: Double
        if let v = num("energy-kcal_100g") ?? num("energy-kcal") {
            kcal = v
        } else if let kj = num("energy_100g") ?? num("energy") {
            kcal = kj / 4.184
        } else {
            return nil
        }
        guard kcal > 0, kcal <= 900 else { return nil }

        var servings: [Serving] = []
        if let grams = OFFProduct.parseGrams(servingSize) {
            servings.append(Serving(label: servingSize ?? "1 serving", grams: grams))
        }

        return FoodHit(
            id: "off:\(id)",
            name: name,
            brand: brands?.value,
            kcal: kcal,
            protein: num("proteins_100g") ?? num("proteins") ?? 0,
            carbs: num("carbohydrates_100g") ?? num("carbohydrates") ?? 0,
            fat: num("fat_100g") ?? num("fat") ?? 0,
            servings: servings,
            source: .openFoodFacts
        )
    }

    private func num(_ key: String) -> Double? {
        guard let v = nutriments?[key]?.double, v.isFinite, v >= 0 else { return nil }
        return v
    }

    /// "30 g", "1 cup (240ml)", "50g" → grams. nil when there's no gram figure.
    static func parseGrams(_ text: String?) -> Double? {
        guard let text, !text.isEmpty else { return nil }
        let pattern = #"([\d]+(?:[.,]\d+)?)\s*(g|ml)\b"#
        guard let regex = try? NSRegularExpression(pattern: pattern, options: .caseInsensitive),
              let match = regex.firstMatch(in: text, range: NSRange(text.startIndex..., in: text)),
              let range = Range(match.range(at: 1), in: text)
        else { return nil }
        let value = Double(text[range].replacingOccurrences(of: ",", with: "."))
        guard let value, value > 0, value <= 2000 else { return nil }
        return value
    }
}

/// Nutriment values arrive as numbers or numeric strings depending on the field.
private struct OFFValue: Decodable {
    let double: Double?

    init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()
        if let d = try? container.decode(Double.self) {
            double = d
        } else if let s = try? container.decode(String.self) {
            double = Double(s)
        } else {
            double = nil
        }
    }
}

private struct ProductResponse: Decodable {
    let status: Int
    let product: OFFProduct?
}

private struct SearchResponse: Decodable {
    let hits: [OFFProduct]
}

File: web/foodsearch.js
/* Foodie — food lookup.
 *
 * Two sources, both free and key-less:
 *   1. foods.js — a bundled table of common whole foods. Instant and offline.
 *   2. Open Food Facts — branded products and barcodes.
 *
 * Only the CORS-enabled Open Food Facts hosts are used here. The faster
 * search.openfoodfacts.org endpoint sends no Access-Control-Allow-Origin, so
 * browsers can't reach it; the legacy cgi/search.pl can, but sheds load with a
 * 503 under pressure, hence the retry below. Barcode lookups use the v2 product
 * endpoint, which is both CORS-enabled and reliable.
 *
 * Everything is normalised to per-100 g macros so one portion calculation
 * serves both sources.
 */
window.FoodieSearch = (function () {
  "use strict";

  var SEARCH_URL = "https://world.openfoodfacts.org/cgi/search.pl";
  var PRODUCT_URL = "https://world.openfoodfacts.org/api/v2/product/";
  var FIELDS = "code,product_name,brands,nutriments,serving_size";
  var TIMEOUT_MS = 12000;

  // Barcodes never change their nutrition, so remember them for this page load.
  var barcodeCache = {};

  // ---- Bundled table --------------------------------------------------------

  // Rows: [id, name, kcal, p, c, f, servings]
  var COMMON = (window.FOODIE_COMMON_FOODS || []).map(function (r) {
    return {
      id: "common:" + r[0],
      name: r[1],
      brand: "",
      kcal: r[2], p: r[3], c: r[4], f: r[5],
      servings: (r[6] || []).map(function (s) { return { label: s[0], grams: s[1] }; }),
      source: "common",
    };
  });

  function normalize(s) {
    return String(s).toLowerCase().normalize("NFD").replace(/[\u0300-\u036f]/g, "");
  }

  /* Higher is better, 0 means no match. Whole-name prefixes beat word starts,
     so "chick" ranks "Chicken breast" above "Chickpeas". Shorter names win ties. */
  function score(name, q) {
    var n = normalize(name);
    var i = n.indexOf(q);
    if (i < 0) return 0;
    var base = i === 0 ? 100 : (/[\s,\-]/.test(n.charAt(i - 1)) ? 60 : 20);
    return base - Math.floor(n.length / 4);
  }

  function searchCommon(query, limit) {
    var q = normalize(String(query).trim());
    if (q.length < 2) return [];
    var scored = [];
    for (var i = 0; i < COMMON.length; i++) {
      var s = score(COMMON[i].name, q);
      if (s > 0) scored.push({ food: COMMON[i], s: s, i: i });
    }
    scored.sort(function (a, b) { return b.s - a.s || a.i - b.i; });
    return scored.slice(0, limit || 8).map(function (x) { return x.food; });
  }

  // ---- Open Food Facts ------------------------------------------------------

  function fetchJson(url) {
    // AbortController keeps a stalled request from hanging the search forever.
    var ctrl = typeof AbortController !== "undefined" ? new AbortController() : null;
    var timer = setTimeout(function () { if (ctrl) ctrl.abort(); }, TIMEOUT_MS);
    return fetch(url, {
      signal: ctrl ? ctrl.signal : undefined,
      headers: { Accept: "application/json" },
    }).then(function (res) {
      clearTimeout(timer);
      if (!res.ok) {
        var err = new Error("http " + res.status);
        err.status = res.status;
        throw err;
      }
      return res.json();
    }, function (e) {
      clearTimeout(timer);
      throw e;
    });
  }

  function num(nutriments, key) {
    if (!nutriments) return null;
    var v = nutriments[key];
    if (typeof v === "string") v = parseFloat(v);
    if (typeof v !== "number" || !isFinite(v) || v < 0) return null;
    return v;
  }

  /* "30 g", "1 cup (240ml)", "50g" -> grams. null when there's no gram figure. */
  function parseGrams(text) {
    if (!text) return null;
    var m = String(text).match(/(\d+(?:[.,]\d+)?)\s*(g|ml)\b/i);
    if (!m) return null;
    var v = parseFloat(m[1].replace(",", "."));
    return isFinite(v) && v > 0 && v <= 2000 ? v : null;
  }

  /* `brands` is a comma-joined string on the product endpoint and an array on
     the search endpoint. Take the first name either way. */
  function firstBrand(brands) {
    if (Array.isArray(brands)) return brands.length ? String(brands[0]).trim() : "";
    if (typeof brands === "string") return brands.split(",")[0].trim();
    return "";
  }

  /* Normalise one OFF product, or null when it has no usable calorie figure. */
  function toHit(p, codeOverride) {
    if (!p) return null;
    var name = String(p.product_name || "").trim();
    if (!name) return null;

    var n = p.nutriments;
    var kcal = num(n, "energy-kcal_100g");
    if (kcal === null) kcal = num(n, "energy-kcal");
    if (kcal === null) {
      var kj = num(n, "energy_100g");
      if (kj === null) kj = num(n, "energy");
      if (kj === null) return null;
      kcal = kj / 4.184;
    }
    // Nothing edible exceeds ~900 kcal/100 g (pure fat); above that the entry is bad data.
    if (!(kcal > 0) || kcal > 900) return null;

    var servings = [];
    var grams = parseGrams(p.serving_size);
    if (grams) servings.push({ label: String(p.serving_size).trim(), grams: grams });

    return {
      id: "off:" + (codeOverride || p.code || name),
      name: name,
      brand: firstBrand(p.brands),
      kcal: kcal,
      p: num(n, "proteins_100g") || num(n, "proteins") || 0,
      c: num(n, "carbohydrates_100g") || num(n, "carbohydrates") || 0,
      f: num(n, "fat_100g") || num(n, "fat") || 0,
      servings: servings,
      source: "off",
    };
  }

  /* Branded matches. Resolves to [] on any failure — bundled results still show. */
  function searchProducts(query, limit) {
    var q = String(query).trim();
    if (q.length < 2) return Promise.resolve([]);
    var url = SEARCH_URL +
      "?search_terms=" + encodeURIComponent(q) +
      "&search_simple=1&action=process&json=1" +
      "&page_size=" + (limit || 20) +
      "&fields=" + FIELDS;

    function attempt(retriesLeft) {
      return fetchJson(url).catch(function (e) {
        // 503 means the shared instance is shedding load; one retry usually lands.
        if (retriesLeft > 0 && e.status === 503) {
          return new Promise(function (r) { setTimeout(r, 600); }).then(function () {
            return attempt(retriesLeft - 1);
          });
        }
        throw e;
      });
    }

    return attempt(1).then(function (data) {
      var products = (data && data.products) || [];
      var out = [];
      // OFF holds one entry per country/packaging, so a popular product comes
      // back a dozen times. Collapse rows that agree on brand, name and kcal.
      var seen = {};
      for (var i = 0; i < products.length; i++) {
        var hit = toHit(products[i]);
        if (!hit) continue;
        var key = normalize(hit.brand + "|" + hit.name) + "|" + Math.round(hit.kcal);
        if (seen[key]) continue;
        seen[key] = true;
        out.push(hit);
      }
      return out;
    }, function () {
      return [];
    });
  }

  /* Barcode lookup. Rejects with a user-facing `.message` so the UI can show it. */
  function lookupBarcode(code) {
    var c = String(code).trim();
    if (!/^\d{6,14}$/.test(c)) {
      return Promise.reject(new Error("That doesn't look like a barcode."));
    }
    if (barcodeCache[c]) return Promise.resolve(barcodeCache[c]);

    return fetchJson(PRODUCT_URL + encodeURIComponent(c) + ".json?fields=" + FIELDS)
      .catch(function (e) {
        if (e.status === 404) throw new Error("No product found for that barcode.");
        throw new Error(
          navigator.onLine
            ? "Food database is busy. Try again in a moment."
            : "You're offline — bundled foods still work."
        );
      })
      .then(function (data) {
        if (!data || data.status !== 1 || !data.product) {
          throw new Error("No product found for that barcode.");
        }
        var hit = toHit(data.product, c);
        if (!hit) throw new Error("That product has no nutrition data yet.");
        barcodeCache[c] = hit;
        return hit;
      });
  }

  // ---- Portions -------------------------------------------------------------

  /* Portions offered in the picker: the food's own servings, then 100 g. */
  function portionOptions(hit) {
    return (hit.servings || []).concat([{ label: "100 g", grams: 100 }]);
  }

  /* Picker text for a serving. Open Food Facts labels often already spell out
     the weight ("1 portion (40 g)"), so only append it when it's missing. */
  function portionLabel(serving) {
    if (/\d\s*g\b/i.test(serving.label)) return serving.label;
    return serving.label + " (" + Math.round(serving.grams) + " g)";
  }

  /* Scale a hit to a gram weight and round into the integer log-entry shape. */
  function entryFor(hit, grams) {
    var k = grams / 100;
    var title = hit.brand ? hit.brand + " " + hit.name : hit.name;
    return {
      name: title,
      kcal: Math.round(hit.kcal * k),
      p: Math.round(hit.p * k),
      c: Math.round(hit.c * k),
      f: Math.round(hit.f * k),
    };
  }

  return {
    searchCommon: searchCommon,
    searchProducts: searchProducts,
    lookupBarcode: lookupBarcode,
    portionOptions: portionOptions,
    portionLabel: portionLabel,
    entryFor: entryFor,
  };
})();
